### PR TITLE
[Kubeapps Chart] Simplify ingress configuration for cert-manager

### PR DESCRIPTION
**Pull request by juan131**
### What this PR does / why we need it:

This PR allows the user to configure the proper annotations when using cert-manager to handle TLS certificates in an easy way. The user just needs to setup a boolean to true to do so.

E.g.:

```
helm install --name kubeapps --namespace kubeapps bitnami/kubeapps \
  --set ingress.enabled=true \
  --set ingress.hosts[0].name=kubeapps.custom.domain \
  --set ingress.hosts[0].tls=true \
  --set ingress.hosts[0].tlsSecret=kubeapps-tls \
  --set ingress.hosts[0].certManager=true
```

This will generate the following ingress:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: kubeapps
  namespace: kubeapps
  labels:
    app: kubeapps
    chart: kubeapps-0.8.0
    release: kubeapps
    heritage: Tiller
  annotations:
    kubernetes.io/tls-acme: "true"
spec:
  rules:
  - host: kubeapps.custom.domain
    http:
      paths:
      - path: /
        backend:
          serviceName: kubeapps
          servicePort: http
  tls:
  - hosts:
    - kubeapps.custom.domain 
    secretName: kubeapps-tls
```

See also https://github.com/bitnami/charts/pull/882

### CODE REVIEW

File: chart/kubeapps/Chart.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: kubeapps
-version: 0.7.0
-appVersion: DEVEL
+version: 0.8.0
+appVersion: v1.0.0-beta.2
```
Comment (andresmgot) [inline chart/kubeapps/Chart.yaml]: revert this change, here it should be `DEVEL`

File: chart/kubeapps/values.yaml [hunk 2]
```diff
@@ -60,7 +90,7 @@ apprepository:
   image:
     registry: docker.io
     repository: kubeapps/apprepository-controller
-    tag: latest
+    tag: v1.0.0-beta.2
```
Comment (andresmgot) [inline chart/kubeapps/values.yaml]: the same here, it's needed for this to be `latest`

File: chart/kubeapps/templates/ingress.yaml [hunk 3]
```diff
@@ -1,38 +1,36 @@
 {{- if .Values.ingress.enabled -}}
-{{- $fullName := include "kubeapps.fullname" . -}}
-{{- $ingressPath := .Values.ingress.path -}}
+{{- range .Values.ingress.hosts }}
```
Comment (migmartri) [inline chart/kubeapps/templates/ingress.yaml]: so this is creating an ingress per host? Why don't we want to use the multi-host capabilities of a single ingress?

Having different ingresses could mean that cert-manager instead of issuing a cert with multiple CN will potentially do different certs no?

File: chart/kubeapps/values.yaml [hunk 4]
```diff
@@ -6,16 +6,46 @@
...
+
+    # Ingress annotations done as key:value pairs
+    # For a full list of possible ingress annotations, please see
+    # ref: https://github.com/kubernetes/ingress-nginx/blob/master/docs/annotations.md
+    #
+    # If certManager is set to true, annotation kubernetes.io/tls-acme: "true" will automatically be set
+    annotations:
+    #  kubernetes.io/ingress.class: nginx
+
+  secrets:
```
Comment (migmartri) [inline chart/kubeapps/values.yaml]: I can not see the reference where this is being used.

File: chart/kubeapps/README.md [hunk 5]
```diff
@@ -171,15 +171,17 @@ To enable ingress integration, please set `ingress.enabled` to `true`
...
 
 ##### TLS
 
-TLS can be configured using the `ingress.tls` object in the same format that the Kubernetes Ingress requests. Please see [this example](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx/examples/tls) for more information.
+TLS can be configured using setting the `ingress.hosts[].tls` boolean of the corresponding hostname to true, then you can choose the TLS secret name setting `ingress.hosts[].tlsSecret`. Please see [this example](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx/examples/tls) for more information.
+
+If your cluster has a [cert-manager](https://github.com/jetstack/cert-manager) add-on to automate the management and issuance of TLS certificates, set `ingress.hosts[].certManager` boolean to true to enable the corresponding annotations for cert-manager. Otherwise, you can provide your own certificates using the `ingress.secrets` object.
```
Comment (migmartri) [inline chart/kubeapps/README.md]: Maybe you can add here the example you put in the description

```
helm install --name kubeapps --namespace kubeapps bitnami/kubeapps \
  --set ingress.enabled=true \
  --set ingress.hosts[0].name=kubeapps.custom.domain \
  --set ingress.hosts[0].tls=true \
  --set ingress.hosts[0].tlsSecret=kubeapps-tls \
  --set ingress.hosts[0].certManager=true
```

**Comment by migmartri** [comment]
Out of curiosity, could you also post what would have been the equivalent to the following command before the change? 
```
helm install --name kubeapps --namespace kubeapps bitnami/kubeapps \
  --set ingress.enabled=true \
  --set ingress.hosts[0].name=kubeapps.custom.domain \
  --set ingress.hosts[0].tls=true \
  --set ingress.hosts[0].tlsSecret=kubeapps-tls \
  --set ingress.hosts[0].certManager=true
```

I am wondering if it would make sense to just add the `certManager` annotation helper but not the multi-host <-> multi-ingress support.

Thanks!

File: chart/kubeapps/README.md (hunk 5 above)
Comment (juan131) [reply chart/kubeapps/README.md]: 👍

File: chart/kubeapps/Chart.yaml (hunk 1 above)
Comment (juan131) [reply chart/kubeapps/Chart.yaml]: 👍

File: chart/kubeapps/values.yaml (hunk 2 above)
Comment (juan131) [reply chart/kubeapps/values.yaml]: 👍

File: chart/kubeapps/templates/ingress.yaml (hunk 3 above)
Comment (juan131) [reply chart/kubeapps/templates/ingress.yaml]: Yes, I think it's better to use `multi-host` since it requires issuing many certificates

**Comment by juan131** [comment]
>  In the example above, you referenced --set ingress.hosts[0].certManager=true which seems to reference to host[0] while the annotation is set at the ingress top level. Shouldn't then not be scoped inside the host?

Totally agree!! We should set it at the ingress top level.

> I am assuming that kubeapps.juan.bkpr.run is a typo and was actually meant to be kubeapps.custom.domain?

Yes, it's a typo.

> I am wondering if it would make sense to just add the certManager annotation helper but not the multi-host <-> multi-ingress support.

I'd rather user `multi-host` too, let me adapt it!

File: chart/kubeapps/templates/ingress.yaml [hunk 6]
```diff
@@ -1,38 +1,38 @@
...
 metadata:
-  name: {{ $fullName }}
+  name: {{ template "kubeapps.fullname" $ }}
   labels:
-    app: {{ include "kubeapps.name" . }}
-    chart: {{ include "kubeapps.chart" . }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
-{{- with .Values.ingress.annotations }}
+    app: {{ template "kubeapps.name" $ }}
```
Comment (migmartri) [inline chart/kubeapps/templates/ingress.yaml]: Thanks for reverting back to template since it seems that here was the only place in which we used `include`.

Comment (migmartri) [reply chart/kubeapps/templates/ingress.yaml]: Why are you using `$` instead of current scope `.`? It seems that in the other templates we use `.`

File: chart/kubeapps/templates/ingress.yaml [hunk 7]
```diff
@@ -1,38 +1,38 @@
...
+    chart: {{ template "kubeapps.chart" $ }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
   annotations:
```
Comment (migmartri) [inline chart/kubeapps/templates/ingress.yaml]: note that this key `annotations` in this new case will be always added regardless of we passing `.Values.ingress.annotations` or  `.Values.ingress.certManager ` is that ok?

File: chart/kubeapps/templates/ingress.yaml [hunk 8]
```diff
@@ -1,38 +1,38 @@
...
-{{ toYaml . | indent 4 }}
-{{- end }}
+    {{- if .Values.ingress.certManager }}
+    kubernetes.io/tls-acme: "true"
+    {{- end }}
+    {{- range $key, $value := .Values.ingress.annotations }}
+    {{ $key }}: {{ $value | quote }}
```
Comment (migmartri) [inline chart/kubeapps/templates/ingress.yaml]: Why do you prefer this approach vs dumping everything like before? `{{ toYaml . | indent 4 }}`

File: chart/kubeapps/templates/ingress.yaml [hunk 9]
```diff
@@ -1,38 +1,38 @@
...
+  {{- end }}
+  tls:
+  {{- range .Values.ingress.hosts }}
+  {{- if .tls }}
+  - hosts:
+    - {{ .name }}
+    secretName: {{ .tlsSecret }}
+  {{- end }}
   {{- end }}
+---
```
Comment (migmartri) [inline chart/kubeapps/templates/ingress.yaml]: is this multi-yaml new line neccessary?

File: chart/kubeapps/templates/tls-secrets.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.ingress.enabled }}
+{{- range .Values.ingress.secrets }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ template "kubeapps.fullname" $ }}
```
Comment (migmartri) [inline chart/kubeapps/templates/tls-secrets.yaml]: isn't the name going to collide if there are two secrets?

File: chart/kubeapps/templates/ingress.yaml (hunk 6 above)
Comment (juan131) [reply chart/kubeapps/templates/ingress.yaml]: We can use `.`. I'll update it. The `$` was necessary when we were using `{{- range ...` at the top.

File: chart/kubeapps/templates/ingress.yaml (hunk 8 above)
Comment (juan131) [reply chart/kubeapps/templates/ingress.yaml]: I just used because it's how we use it on others charts. Anyway, it's only adding the value of quoting the value...

I think it's irrelevant to use one or the another.

File: chart/kubeapps/templates/ingress.yaml (hunk 9 above)
Comment (juan131) [reply chart/kubeapps/templates/ingress.yaml]: No

File: chart/kubeapps/templates/tls-secrets.yaml (hunk 10 above)
Comment (juan131) [reply chart/kubeapps/templates/tls-secrets.yaml]: Nice catch!! I guess we can do sth like:
```
metadata:
  name: {{ .name }}
```

File: chart/kubeapps/templates/ingress.yaml (hunk 7 above)
Comment (juan131) [reply chart/kubeapps/templates/ingress.yaml]: It's fine. The result is: no annotations

File: chart/kubeapps/templates/ingress.yaml (hunk 8 above)
Comment (migmartri) [reply chart/kubeapps/templates/ingress.yaml]: > I just used because it's how we use it on others charts

ok, it makes sense. thanks for the explanation

We might need to move the rest of the chart to this approach in another PR since I believe that we use `{{ toYaml . | indent 4 }}` in other places.

**Comment by migmartri** [comment]
BTW, now that I think about this. I am a little bit worried about the change on the way we define the hosts since it will mean a breaking change right? Meaning that if somebody today has an automated upgrade process that relies on having ingress enable it will fail right? 

In that case, if we still decide to go ahead with the change we might need to bump the mayor version not the minor one.

what do you think @prydonius @juan131 @andresmgot ?

**Comment by juan131** [comment]
I agree. It breaks backwards compatibility and, therefore, it should be considered a major change.

**Comment by juan131** [comment]
PR updated bumping the major version.

**Comment by juan131** [comment]
IMHO doing a major release of a chart should not be a drama. 

We can add the steps you need to perform to do an upgrade if you consider it necessary but I think it's something normal since charts are improved constantly and some of these changes can break backward compatibility.

**Comment by migmartri** [comment]
Actually @andresmgot raised a good point that in semver, for 0.x versions, minor version changes are considered breaking. So feel free to revert back to 0.9 (in your case since 0.8 has already been taken).

Thanks @juan131!
